File: leetcode75/7-238. Product of Array Except Self/main.rs
pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
    let mut product_array = vec![1; nums.len()];
    let mut prefix_product = 1;

    for (i, &v) in nums.iter().enumerate() {
        if i == 0 {
            continue;
        }
        product_array[i] = prefix_product * nums[i - 1];
        prefix_product = product_array[i];
    }

    let mut postfix_product = 1;

    for (i, v) in nums.iter().enumerate().rev() {
        product_array[i] = postfix_product * product_array[i];
        postfix_product = postfix_product * nums[i];
    }

    return product_array;
}

fn main() {
    println!("{:?}", product_except_self(vec![1, 2, 3, 4]));
}

File: leetcode75/7-238. Product of Array Except Self/main.py
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        product_array = [1] * len(nums)
        prefix_product = 1
        postfix_product = 1

        # [1,2,3,4]
        # each value will in product_array will have the product of
        # all values in left
        for i in range(1, len(nums)):
            product = prefix_product * nums[i - 1]
            product_array[i] = product
            prefix_product = product

        for i in range(len(nums) - 1, -1, -1):
            print(i)
            product_array[i] = postfix_product * product_array[i]
            postfix_product = nums[i] * postfix_product
        return product_array


s = Solution()
print(s.productExceptSelf([1, 2, 3, 4]))
